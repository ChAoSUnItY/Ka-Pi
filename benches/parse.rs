use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ka_pi::parse::to_class;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse main", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    )))));

    c.bench_function("parse enum", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Enum.class"
    )))));

    c.bench_function("parse record", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Record.class"
    )))));

    c.bench_function("parse visible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/VisibleAnnotation.class"
    )))));

    c.bench_function("parse invisible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/InvisibleAnnotation.class"
    )))));

    c.bench_function("parse annotation target", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/AnnotationTarget.class"
    )))));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
