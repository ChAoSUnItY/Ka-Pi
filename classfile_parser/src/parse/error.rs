use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match opcode but got opcode value {0}")]
    MatchOutOfBoundOpcode(u8),
    #[error("attempt to match sub opcode in wide opcode but got {0}, while {1:?} is allowed as sub opcode of wide opcode")]
    MatchOutOfBoundWideOpcode(u8, Vec<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
