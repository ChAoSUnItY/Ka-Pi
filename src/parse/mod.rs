use std::io::Read;

pub use error::{ParseError, ParseResult};
pub use signature::{class_signature, field_signature, method_signature};

use crate::node::class::Class;

pub(crate) mod attribute;
pub(crate) mod class;
pub(crate) mod constant;
pub(crate) mod error;
pub(crate) mod field;
pub(crate) mod method;
pub(crate) mod signature;

/// Parses class file based on given options.
pub fn to_class<R: Read>(class_bytes: &mut R, option: ParsingOption) -> ParseResult<Class> {
    let class = class::class(class_bytes, option)?;

    let mut remain = vec![];

    class_bytes.read_to_end(&mut remain)?;

    if !remain.is_empty() {
        Err(ParseError::Remains(remain.len()))
    } else {
        Ok(class)
    }
}

/// Parsing options allows marking some parsing phase optional.
#[derive(Debug, Default)]
pub struct ParsingOption {
    parse_attribute: bool,
}

impl ParsingOption {
    /// Skips on attribute struct parsing.
    pub fn parse_attribute(mut self) -> Self {
        self.parse_attribute = true;

        self
    }
}
