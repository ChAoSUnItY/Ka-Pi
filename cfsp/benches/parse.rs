use cfsp::parse::{to_class, ParsingOption};
use std::fs;
use std::io::Cursor;

use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, Throughput};

fn parsing_classes(c: &mut Criterion) {
    let mut g = c.benchmark_group("parse class");

    let compiled_source_folder =
        fs::read_dir("../compiled_source/out/production/compiled_source").unwrap();

    for entry in compiled_source_folder {
        let entry = entry.as_ref().unwrap().path();
        let class_name = entry.file_name().unwrap().to_str().unwrap();
        let class_bytes = fs::read(&entry).unwrap();

        g.throughput(Throughput::Bytes(class_bytes.len() as u64));
        g.bench_with_input(
            BenchmarkId::from_parameter(class_name),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default().parse_attribute())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );

        g.bench_with_input(
            BenchmarkId::from_parameter(format!("{class_name} (No attribute parsing)")),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }
}

criterion_group!(benches, parsing_classes);
criterion_main!(benches);
