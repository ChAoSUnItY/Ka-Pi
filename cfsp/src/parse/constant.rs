use std::cell::RefCell;
use std::io::Read;

use byteorder::{BigEndian, ReadBytesExt};

use crate::node::constant;
use crate::node::constant::{
    Class, Constant, ConstantPool, Double, Dynamic, FieldRef, Float, Integer, InterfaceMethodRef,
    InvokeDynamic, Long, MethodHandle, MethodRef, MethodType, Module, NameAndType, Package, Utf8,
};
use crate::parse::error::{ParseError, ParseResult};

pub(super) fn constant_pool<R: Read>(input: &mut R) -> ParseResult<(u16, ConstantPool)> {
    let len = input.read_u16::<BigEndian>()?;
    let mut constant_pool = ConstantPool::with_capacity(len);
    let mut constant_counter = 0;

    while constant_counter < len - 1 {
        let constant = constant(input)?;

        if constant.occupies_2_slots() {
            constant_counter += 2;
        } else {
            constant_counter += 1;
        }

        constant_pool.add(constant);
    }

    Ok((len, constant_pool))
}

fn constant<R: Read>(input: &mut R) -> ParseResult<Constant> {
    let tag = input.read_u8()?;

    let constant = match tag {
        1 => {
            let length = input.read_u16::<BigEndian>()?;
            let mut bytes = vec![0; length as usize];

            input.read_exact(&mut bytes)?;

            Constant::Utf8(Utf8 {
                length,
                bytes,
                string: RefCell::new(None),
            })
        }
        3 => {
            let mut bytes = [0; 4];

            input.read_exact(&mut bytes)?;

            Constant::Integer(Integer { bytes })
        }
        4 => {
            let mut bytes = [0; 4];

            input.read_exact(&mut bytes)?;

            Constant::Float(Float { bytes })
        }
        5 => {
            let mut high_bytes = [0; 4];
            let mut low_bytes = [0; 4];

            input.read_exact(&mut high_bytes)?;
            input.read_exact(&mut low_bytes)?;

            Constant::Long(Long {
                high_bytes,
                low_bytes,
            })
        }
        6 => {
            let mut high_bytes = [0; 4];
            let mut low_bytes = [0; 4];

            input.read_exact(&mut high_bytes)?;
            input.read_exact(&mut low_bytes)?;

            Constant::Double(Double {
                high_bytes,
                low_bytes,
            })
        }
        7 => {
            let name_index = input.read_u16::<BigEndian>()?;

            Constant::Class(Class { name_index })
        }
        8 => {
            let string_index = input.read_u16::<BigEndian>()?;

            Constant::String(constant::String { string_index })
        }
        9 => {
            let class_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::FieldRef(FieldRef {
                class_index,
                name_and_type_index,
            })
        }
        10 => {
            let class_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::MethodRef(MethodRef {
                class_index,
                name_and_type_index,
            })
        }
        11 => {
            let class_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::InterfaceMethodRef(InterfaceMethodRef {
                class_index,
                name_and_type_index,
            })
        }
        12 => {
            let name_index = input.read_u16::<BigEndian>()?;
            let type_index = input.read_u16::<BigEndian>()?;

            Constant::NameAndType(NameAndType {
                name_index,
                type_index,
            })
        }
        15 => {
            let reference_kind = input.read_u8()?;
            let reference_index = input.read_u16::<BigEndian>()?;

            Constant::MethodHandle(MethodHandle {
                reference_kind,
                reference_index,
            })
        }
        16 => {
            let descriptor_index = input.read_u16::<BigEndian>()?;

            Constant::MethodType(MethodType { descriptor_index })
        }
        17 => {
            let bootstrap_method_attr_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::Dynamic(Dynamic {
                bootstrap_method_attr_index,
                name_and_type_index,
            })
        }
        18 => {
            let bootstrap_method_attr_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::InvokeDynamic(InvokeDynamic {
                bootstrap_method_attr_index,
                name_and_type_index,
            })
        }
        19 => {
            let name_index = input.read_u16::<BigEndian>()?;

            Constant::Module(Module { name_index })
        }
        20 => {
            let name_index = input.read_u16::<BigEndian>()?;

            Constant::Package(Package { name_index })
        }
        _ => {
            return Err(ParseError::MatchOutOfBoundUsize(
                "constant tag",
                vec!["1..=20"],
                tag as usize,
            ))
        }
    };

    Ok(constant)
}
