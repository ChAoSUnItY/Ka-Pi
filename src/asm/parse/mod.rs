use std::fs::File;
use std::io::Read;
use std::ops::RangeFrom;
use std::path::Path;

use nom::combinator::map;
use nom::error::ParseError;
use nom::number::complete::be_u16;
use nom::{IResult, InputIter, InputLength, Slice};

use crate::asm::node::access_flag::AccessFlag;
use crate::asm::node::class::Class;
use crate::asm::node::constant::ConstantPool;
use crate::error::{KapiError, KapiResult};

pub mod attribute;
pub mod class;
pub mod constant;
pub mod field;
pub mod method;

pub fn read_class<P: AsRef<Path>>(class_path: P) -> KapiResult<Class> {
    let class_path = class_path.as_ref();
    let mut file = match File::open(class_path) {
        Ok(file) => file,
        Err(err) => {
            return Err(KapiError::ClassParseError(format!(
                "Unable to open class file {}, reason: {}",
                class_path.display(),
                err
            )))
        }
    };
    let mut class_bytes = Vec::new();

    if let Err(err) = file.read_to_end(&mut class_bytes) {
        return Err(KapiError::ClassParseError(format!(
            "Unable to read class file {}, reason: {}",
            class_path.display(),
            err
        )));
    }

    to_class(&class_bytes)
}

pub fn to_class(class_bytes: &[u8]) -> KapiResult<Class> {
    match class::class(&class_bytes[..]) {
        Ok((remain, class)) => {
            if !remain.is_empty() {
                Err(KapiError::ClassParseError(format!("Unable to parse class bytes, reason: class is fully parsed but there are {} bytes left, {remain:?}", remain.len())))
            } else {
                Ok(class)
            }
        }
        Err(err) => Err(KapiError::ClassParseError(format!(
            "Unable parse class bytes, reason: {err}"
        ))),
    }
}

fn collect<I, LP, TP, L, T, E: ParseError<I>>(
    mut len_parser: LP,
    mut item_parser: TP,
) -> impl FnMut(I) -> IResult<I, (L, Vec<T>), E>
where
    I: Slice<RangeFrom<usize>> + InputIter<Item = u8> + InputLength,
    LP: FnMut(I) -> IResult<I, L, E>,
    TP: FnMut(I) -> IResult<I, T, E>,
    L: Into<u64> + Copy,
{
    move |input: I| {
        let (mut input, len) = len_parser(input)?;
        let length = len.into();
        let mut items = Vec::with_capacity(length as usize);

        for _ in 0..length {
            let (remain, item) = item_parser(input)?;

            items.push(item);
            input = remain;
        }

        Ok((input, (len, items)))
    }
}

fn collect_with_constant_pool<'parser, 'constant_pool: 'parser, I, LP, TP, L, T, E: ParseError<I>>(
    mut len_parser: LP,
    mut item_parser: TP,
    constant_pool: &'constant_pool ConstantPool,
) -> impl FnMut(I) -> IResult<I, (L, Vec<T>), E> + '_
where
    I: Slice<RangeFrom<usize>> + InputIter<Item = u8> + InputLength,
    LP: FnMut(I) -> IResult<I, L, E> + 'parser + 'constant_pool,
    TP: FnMut(I, &'constant_pool ConstantPool) -> IResult<I, T, E> + 'parser + 'constant_pool,
    L: Into<u64> + Copy,
{
    move |input: I| {
        let (mut input, len) = len_parser(input)?;
        let length = len.into();
        let mut items = Vec::with_capacity(length as usize);

        for _ in 0..length {
            let (remain, item) = item_parser(input, constant_pool)?;

            items.push(item);
            input = remain;
        }

        Ok((input, (len, items)))
    }
}

fn access_flag<F>(input: &[u8]) -> IResult<&[u8], Vec<F>>
where
    F: AccessFlag,
{
    map(be_u16, F::mask_access_flags)(input)
}
