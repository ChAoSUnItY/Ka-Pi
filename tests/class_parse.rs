use insta::assert_yaml_snapshot;
use ka_pi::parse::{to_class, ParsingOption};
use std::io::Cursor;

use ka_pi::parse::ParseResult;

#[test]
fn test_main() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_enum() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/Enum.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_record() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/Record.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_visible_annotation() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/VisibleAnnotation.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_invisible_annotation() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/InvisibleAnnotation.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_annotation_target() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/AnnotationTarget.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}
