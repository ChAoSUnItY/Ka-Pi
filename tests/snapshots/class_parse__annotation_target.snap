---
source: tests/class_parse.rs
expression: "to_class(include_bytes!(\"../compiled_source/out/production/compiled_source/AnnotationTarget.class\"))?"
---
magic_number:
  - start: 0
    end: 4
  - - 202
    - 254
    - 186
    - 190
java_version:
  - start: 4
    end: 8
  - V17
constant_pool_count:
  - start: 8
    end: 10
  - 19
constant_pool:
  - start: 10
    end: 266
  - len: 19
    entries:
      1:
        - start: 10
          end: 15
        - tag:
            - start: 10
              end: 11
            - MethodRef
          constant:
            - start: 11
              end: 15
            - MethodRef:
                class_index:
                  - start: 11
                    end: 13
                  - 2
                name_and_type_index:
                  - start: 13
                    end: 15
                  - 3
      2:
        - start: 15
          end: 18
        - tag:
            - start: 15
              end: 16
            - Class
          constant:
            - start: 16
              end: 18
            - Class:
                name_index:
                  - start: 16
                    end: 18
                  - 4
      3:
        - start: 18
          end: 23
        - tag:
            - start: 18
              end: 19
            - NameAndType
          constant:
            - start: 19
              end: 23
            - NameAndType:
                name_index:
                  - start: 19
                    end: 21
                  - 5
                type_index:
                  - start: 21
                    end: 23
                  - 6
      4:
        - start: 23
          end: 42
        - tag:
            - start: 23
              end: 24
            - Utf8
          constant:
            - start: 24
              end: 42
            - Utf8:
                length:
                  - start: 24
                    end: 26
                  - 16
                bytes:
                  - start: 26
                    end: 42
                  - - 106
                    - 97
                    - 118
                    - 97
                    - 47
                    - 108
                    - 97
                    - 110
                    - 103
                    - 47
                    - 79
                    - 98
                    - 106
                    - 101
                    - 99
                    - 116
      5:
        - start: 42
          end: 51
        - tag:
            - start: 42
              end: 43
            - Utf8
          constant:
            - start: 43
              end: 51
            - Utf8:
                length:
                  - start: 43
                    end: 45
                  - 6
                bytes:
                  - start: 45
                    end: 51
                  - - 60
                    - 105
                    - 110
                    - 105
                    - 116
                    - 62
      6:
        - start: 51
          end: 57
        - tag:
            - start: 51
              end: 52
            - Utf8
          constant:
            - start: 52
              end: 57
            - Utf8:
                length:
                  - start: 52
                    end: 54
                  - 3
                bytes:
                  - start: 54
                    end: 57
                  - - 40
                    - 41
                    - 86
      7:
        - start: 57
          end: 60
        - tag:
            - start: 57
              end: 58
            - Class
          constant:
            - start: 58
              end: 60
            - Class:
                name_index:
                  - start: 58
                    end: 60
                  - 8
      8:
        - start: 60
          end: 79
        - tag:
            - start: 60
              end: 61
            - Utf8
          constant:
            - start: 61
              end: 79
            - Utf8:
                length:
                  - start: 61
                    end: 63
                  - 16
                bytes:
                  - start: 63
                    end: 79
                  - - 65
                    - 110
                    - 110
                    - 111
                    - 116
                    - 97
                    - 116
                    - 105
                    - 111
                    - 110
                    - 84
                    - 97
                    - 114
                    - 103
                    - 101
                    - 116
      9:
        - start: 79
          end: 86
        - tag:
            - start: 79
              end: 80
            - Utf8
          constant:
            - start: 80
              end: 86
            - Utf8:
                length:
                  - start: 80
                    end: 82
                  - 4
                bytes:
                  - start: 82
                    end: 86
                  - - 67
                    - 111
                    - 100
                    - 101
      10:
        - start: 86
          end: 104
        - tag:
            - start: 86
              end: 87
            - Utf8
          constant:
            - start: 87
              end: 104
            - Utf8:
                length:
                  - start: 87
                    end: 89
                  - 15
                bytes:
                  - start: 89
                    end: 104
                  - - 76
                    - 105
                    - 110
                    - 101
                    - 78
                    - 117
                    - 109
                    - 98
                    - 101
                    - 114
                    - 84
                    - 97
                    - 98
                    - 108
                    - 101
      11:
        - start: 104
          end: 125
        - tag:
            - start: 104
              end: 105
            - Utf8
          constant:
            - start: 105
              end: 125
            - Utf8:
                length:
                  - start: 105
                    end: 107
                  - 18
                bytes:
                  - start: 107
                    end: 125
                  - - 76
                    - 111
                    - 99
                    - 97
                    - 108
                    - 86
                    - 97
                    - 114
                    - 105
                    - 97
                    - 98
                    - 108
                    - 101
                    - 84
                    - 97
                    - 98
                    - 108
                    - 101
      12:
        - start: 125
          end: 132
        - tag:
            - start: 125
              end: 126
            - Utf8
          constant:
            - start: 126
              end: 132
            - Utf8:
                length:
                  - start: 126
                    end: 128
                  - 4
                bytes:
                  - start: 128
                    end: 132
                  - - 116
                    - 104
                    - 105
                    - 115
      13:
        - start: 132
          end: 153
        - tag:
            - start: 132
              end: 133
            - Utf8
          constant:
            - start: 133
              end: 153
            - Utf8:
                length:
                  - start: 133
                    end: 135
                  - 18
                bytes:
                  - start: 135
                    end: 153
                  - - 76
                    - 65
                    - 110
                    - 110
                    - 111
                    - 116
                    - 97
                    - 116
                    - 105
                    - 111
                    - 110
                    - 84
                    - 97
                    - 114
                    - 103
                    - 101
                    - 116
                    - 59
      14:
        - start: 153
          end: 166
        - tag:
            - start: 153
              end: 154
            - Utf8
          constant:
            - start: 154
              end: 166
            - Utf8:
                length:
                  - start: 154
                    end: 156
                  - 10
                bytes:
                  - start: 156
                    end: 166
                  - - 83
                    - 111
                    - 117
                    - 114
                    - 99
                    - 101
                    - 70
                    - 105
                    - 108
                    - 101
      15:
        - start: 166
          end: 190
        - tag:
            - start: 166
              end: 167
            - Utf8
          constant:
            - start: 167
              end: 190
            - Utf8:
                length:
                  - start: 167
                    end: 169
                  - 21
                bytes:
                  - start: 169
                    end: 190
                  - - 65
                    - 110
                    - 110
                    - 111
                    - 116
                    - 97
                    - 116
                    - 105
                    - 111
                    - 110
                    - 84
                    - 97
                    - 114
                    - 103
                    - 101
                    - 116
                    - 46
                    - 106
                    - 97
                    - 118
                    - 97
      16:
        - start: 190
          end: 220
        - tag:
            - start: 190
              end: 191
            - Utf8
          constant:
            - start: 191
              end: 220
            - Utf8:
                length:
                  - start: 191
                    end: 193
                  - 27
                bytes:
                  - start: 193
                    end: 220
                  - - 82
                    - 117
                    - 110
                    - 116
                    - 105
                    - 109
                    - 101
                    - 73
                    - 110
                    - 118
                    - 105
                    - 115
                    - 105
                    - 98
                    - 108
                    - 101
                    - 65
                    - 110
                    - 110
                    - 111
                    - 116
                    - 97
                    - 116
                    - 105
                    - 111
                    - 110
                    - 115
      17:
        - start: 220
          end: 242
        - tag:
            - start: 220
              end: 221
            - Utf8
          constant:
            - start: 221
              end: 242
            - Utf8:
                length:
                  - start: 221
                    end: 223
                  - 19
                bytes:
                  - start: 223
                    end: 242
                  - - 76
                    - 86
                    - 105
                    - 115
                    - 105
                    - 98
                    - 108
                    - 101
                    - 65
                    - 110
                    - 110
                    - 111
                    - 116
                    - 97
                    - 116
                    - 105
                    - 111
                    - 110
                    - 59
      18:
        - start: 242
          end: 266
        - tag:
            - start: 242
              end: 243
            - Utf8
          constant:
            - start: 243
              end: 266
            - Utf8:
                length:
                  - start: 243
                    end: 245
                  - 21
                bytes:
                  - start: 245
                    end: 266
                  - - 76
                    - 73
                    - 110
                    - 118
                    - 105
                    - 115
                    - 105
                    - 98
                    - 108
                    - 101
                    - 65
                    - 110
                    - 110
                    - 111
                    - 116
                    - 97
                    - 116
                    - 105
                    - 111
                    - 110
                    - 59
access_flags:
  - start: 266
    end: 268
  - - Public
    - Super
this_class:
  - start: 268
    end: 270
  - 7
super_class:
  - start: 270
    end: 272
  - 2
interfaces_count:
  - start: 272
    end: 274
  - 0
interfaces:
  - start: 274
    end: 274
  - []
fields_count:
  - start: 274
    end: 276
  - 0
fields:
  - start: 276
    end: 276
  - []
methods_count:
  - start: 276
    end: 278
  - 1
methods:
  - start: 278
    end: 339
  - - - start: 278
        end: 339
      - access_flags:
          - start: 278
            end: 280
          - - Public
        name_index:
          - start: 280
            end: 282
          - 5
        descriptor_index:
          - start: 282
            end: 284
          - 6
        attribute_infos_len:
          - start: 284
            end: 286
          - 1
        attribute_infos:
          - start: 286
            end: 339
          - - - start: 286
                end: 339
              - attribute_name_index:
                  - start: 286
                    end: 288
                  - 9
                attribute_len:
                  - start: 288
                    end: 292
                  - 47
                info:
                  - start: 292
                    end: 339
                  - - 0
                    - 1
                    - 0
                    - 1
                    - 0
                    - 0
                    - 0
                    - 5
                    - 42
                    - 183
                    - 0
                    - 1
                    - 177
                    - 0
                    - 0
                    - 0
                    - 2
                    - 0
                    - 10
                    - 0
                    - 0
                    - 0
                    - 6
                    - 0
                    - 1
                    - 0
                    - 0
                    - 0
                    - 3
                    - 0
                    - 11
                    - 0
                    - 0
                    - 0
                    - 12
                    - 0
                    - 1
                    - 0
                    - 0
                    - 0
                    - 5
                    - 0
                    - 12
                    - 0
                    - 13
                    - 0
                    - 0
                attribute:
                  - start: 292
                    end: 339
                  - Code:
                      max_stack:
                        - start: 292
                          end: 294
                        - 1
                      max_locals:
                        - start: 294
                          end: 296
                        - 1
                      code_length:
                        - start: 296
                          end: 300
                        - 5
                      code:
                        - start: 300
                          end: 305
                        - - 42
                          - 183
                          - 0
                          - 1
                          - 177
                      instructions:
                        - start: 300
                          end: 305
                        - - - start: 300
                              end: 301
                            - ALOAD_0
                          - - start: 301
                              end: 304
                            - INVOKESPECIAL:
                                index:
                                  - start: 302
                                    end: 304
                                  - 1
                          - - start: 304
                              end: 305
                            - RETURN
                      exception_table_length:
                        - start: 305
                          end: 307
                        - 0
                      exception_table:
                        - start: 307
                          end: 307
                        - []
                      attributes_length:
                        - start: 307
                          end: 309
                        - 2
                      attributes:
                        - start: 309
                          end: 339
                        - - - start: 309
                              end: 321
                            - attribute_name_index:
                                - start: 309
                                  end: 311
                                - 10
                              attribute_len:
                                - start: 311
                                  end: 315
                                - 6
                              info:
                                - start: 315
                                  end: 321
                                - - 0
                                  - 1
                                  - 0
                                  - 0
                                  - 0
                                  - 3
                              attribute:
                                - start: 315
                                  end: 321
                                - LineNumberTable:
                                    line_number_table_length:
                                      - start: 315
                                        end: 317
                                      - 1
                                    line_number_table:
                                      - start: 317
                                        end: 321
                                      - - - start: 317
                                            end: 321
                                          - start_pc:
                                              - start: 317
                                                end: 319
                                              - 0
                                            line_number:
                                              - start: 319
                                                end: 321
                                              - 3
                          - - start: 321
                              end: 339
                            - attribute_name_index:
                                - start: 321
                                  end: 323
                                - 11
                              attribute_len:
                                - start: 323
                                  end: 327
                                - 12
                              info:
                                - start: 327
                                  end: 339
                                - - 0
                                  - 1
                                  - 0
                                  - 0
                                  - 0
                                  - 5
                                  - 0
                                  - 12
                                  - 0
                                  - 13
                                  - 0
                                  - 0
                              attribute:
                                - start: 327
                                  end: 339
                                - LocalVariableTable:
                                    local_variable_table_length:
                                      - start: 327
                                        end: 329
                                      - 1
                                    local_variable_table:
                                      - start: 329
                                        end: 339
                                      - - - start: 329
                                            end: 339
                                          - start_pc:
                                              - start: 329
                                                end: 331
                                              - 0
                                            length:
                                              - start: 331
                                                end: 333
                                              - 5
                                            name_index:
                                              - start: 333
                                                end: 335
                                              - 12
                                            descriptor_index:
                                              - start: 335
                                                end: 337
                                              - 13
                                            index:
                                              - start: 337
                                                end: 339
                                              - 0
attributes_count:
  - start: 339
    end: 341
  - 2
attributes:
  - start: 341
    end: 365
  - - - start: 341
        end: 349
      - attribute_name_index:
          - start: 341
            end: 343
          - 14
        attribute_len:
          - start: 343
            end: 347
          - 2
        info:
          - start: 347
            end: 349
          - - 0
            - 15
        attribute:
          - start: 347
            end: 349
          - SourceFile:
              source_file_index:
                - start: 347
                  end: 349
                - 15
    - - start: 349
        end: 365
      - attribute_name_index:
          - start: 349
            end: 351
          - 16
        attribute_len:
          - start: 351
            end: 355
          - 10
        info:
          - start: 355
            end: 365
          - - 0
            - 2
            - 0
            - 17
            - 0
            - 0
            - 0
            - 18
            - 0
            - 0
        attribute:
          - start: 355
            end: 365
          - RuntimeInvisibleAnnotations:
              num_annotations:
                - start: 355
                  end: 357
                - 2
              annotations:
                - start: 357
                  end: 365
                - - - start: 357
                      end: 361
                    - type_index:
                        - start: 357
                          end: 359
                        - 17
                      num_element_value_pairs:
                        - start: 359
                          end: 361
                        - 0
                      element_value_pairs:
                        - start: 361
                          end: 361
                        - []
                  - - start: 361
                      end: 365
                    - type_index:
                        - start: 361
                          end: 363
                        - 18
                      num_element_value_pairs:
                        - start: 363
                          end: 365
                        - 0
                      element_value_pairs:
                        - start: 365
                          end: 365
                        - []

